src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs

```diff
@@ -4,114 +4,166 @@
         private IDisposable _workerErrorSubscription;
         private IList<IDisposable> _workerStateSubscriptions = new List<IDisposable>();
-        private ConcurrentDictionary<string, ILanguageWorkerChannel> _channelsDictionary = new ConcurrentDictionary<string, ILanguageWorkerChannel>();
-        private string _language;
+        private ILoggerFactory _loggerFactory;
+        private ScriptJobHostOptions _scriptOptions;
         private bool disposedValue = false;
 
-        public FunctionDispatcher(
-            IScriptEventManager manager,
-            IRpcServer server,
-            CreateChannel channelFactory,
-            IEnumerable<WorkerConfig> workerConfigs,
-            string language)
+        public FunctionDispatcher(IOptions<ScriptJobHostOptions> scriptHostOptions,
+            IMetricsLogger metricsLogger,
+            IScriptEventManager eventManager,
+            ILoggerFactory loggerFactory,
+            IOptions<LanguageWorkerOptions> languageWorkerOptions,
+            ILanguageWorkerChannelManager languageWorkerChannelManager)
         {
-            _eventManager = manager;
-            _server = server;
-            _channelFactory = channelFactory;
-            _language = language;
-            _workerConfigs = workerConfigs ?? throw new ArgumentNullException("workerConfigs");
+            _metricsLogger = metricsLogger;
+            _scriptOptions = scriptHostOptions.Value;
+            _languageWorkerChannelManager = languageWorkerChannelManager;
+            _eventManager = eventManager;
+            _loggerFactory = loggerFactory;
+            _workerConfigs = languageWorkerOptions.Value.WorkerConfigs;
+
             _workerErrorSubscription = _eventManager.OfType<WorkerErrorEvent>()
-                .Subscribe(WorkerError);
+               .Subscribe(WorkerError);
         }
 
-        public IDictionary<WorkerConfig, LanguageWorkerState> LanguageWorkerChannelStates => _channelStates;
+        public IDictionary<string, LanguageWorkerState> LanguageWorkerChannelStates => _channelStates;
 
-        public bool IsSupported(FunctionMetadata functionMetadata)
+        public bool IsSupported(FunctionMetadata functionMetadata, string currentLanguageRuntime)
         {
             if (string.IsNullOrEmpty(functionMetadata.Language))
             {
                 return false;
             }
-            if (string.IsNullOrEmpty(_language))
+            if (string.IsNullOrEmpty(currentLanguageRuntime))
             {
                 return true;
             }
-            return functionMetadata.Language.Equals(_language, StringComparison.OrdinalIgnoreCase);
+            return functionMetadata.Language.Equals(currentLanguageRuntime, StringComparison.OrdinalIgnoreCase);
         }
 
-        internal LanguageWorkerState CreateWorkerState(WorkerConfig config)
+        public LanguageWorkerState CreateWorkerStateWithExistingChannel(string language, ILanguageWorkerChannel languageWorkerChannel)
```

Comment:
Seems like these two methods should be renamed, and be merged into one method with an optional channel param.

Perhaps InitializeChannel? Or just Initialize (initialize for the specified language)